Further code enhancements

sdharar opened the pull request:
Made some significant changes:

I have changed code logic to work with DJI mp4 files. Fixed Unwrap() on None issue for ftype atom. 
Added support for audio track metadata and atom (smhd). Also, I added other missing atoms about the necessary metadata.
Made changes to printing format for parser output, including limiting printing output for stco atoms. Finally, updated signatures for corresponding code files, i

sdharar committed:
Added STCO atom parsing. Parsing works fine but the indentation problem has to be fixed.

sdharar committed:
Made significant changes including: Added new atoms, debugged/fixed code to work with DJI prouced video files,  and other minor changes..

sdharar committed:
Fixed branch merging problem

sdharar committed:
Deleted unnecessary output txt files and vscode settings files

arastoob commented on src/header.rs:
```diff
@@ -5,13 +5,14 @@
     pub name: String,
     pub size: usize,
     pub start: u64,
+    pub end: u64,
```
As we have the `start` and the `size`, the `end` can be calculated: `start + end`. This field and its method is not necessary.

sdharar replied:
Yeah, that's understood but most of the well-known parsers on crates.io print these minimum 4 details in the parsing output. Is there any problem with printing these values?

arastoob replied:
You can print the `end` value, but it does not mean you need a field to keep the `end` value while you can calculate it. Whenever you need to print it: `end = start + size`

sdharar replied:
I noticed one of your older comments on this, the newer commit on the stco-parsing branch removes the end() method and uses the above as suggested.

arastoob commented on src/boxes/mp4_box.rs:
```diff
@@ -17,6 +18,9 @@ pub trait Mp4Box: Downcast {
     fn start(&self) -> u64;
 
+    /// The end address of the box
+    fn end(&self) -> u64;
```
Please see https://github.com/musec/fmvparse/pull/3/files#r872625723

arastoob commented on src/boxes/file_type.rs:
```diff
@@ -17,16 +18,17 @@ pub struct FileType {
         self.header.start
     }
 
+    fn end(&self) -> u64 {
```
Please see https://github.com/musec/fmvparse/pull/3/files#r872625723

arastoob commented on src/boxes/mp4_box.rs:
```diff
@@ -31,6 +35,9 @@ pub trait Mp4Box: Downcast {
     fn level(&self) -> u8;
+
+    /// Print offsets of stco boxes
+    fn offsets(&self) -> Option<Vec<u64>>;
```
If the only module that needs this method is `stco`, this method shouldn't be in the trait, it should be in the only place that is needed, which is the `stco` module and returns a `Vec<u64>` rather than `Option...`

sdharar replied:
I named the method "offset()" just because I didn't think of the right name for the method. The original intention to include this method is to parse additional data further down the atom, apart from atom name and atom size which we already have as part of the logic. For example, ftype may have additional metadata such as major brands, minor versions, and so forth. Similarly, vmhd and smhd have further metadata such as opcolor_r, opcolor_g, opcolor_b, graphics mode, and so forth. Similarly, ChunkOffsetTable is one such metadata in the stco atom that includes offsets. just like how other atoms will have metadata named differently. 

Now one might argue that we might not require this right now, but wouldn't it be good to be prepared if we might need this metadata for example for redaction or conversion process? And I'm confident this will for sure be required and parsed for the conversion process. Having a separate method for each code will then be redundant instead of having one method that works with all atoms?

arastoob replied:
I'm not saying we do not require such methods, they should be in their right place. The `Mp4Box` trait is for defining the common functions for all other modules. If you need more functions in a module, you should have them in the module impl block.

sdharar replied:
This has been resolved based in the newer commit.

arastoob commented on src/boxes/mp4_box.rs:
```diff
@@ -39,10 +46,11 @@ impl std::fmt::Debug for dyn Mp4Box {
+            self.size(),
             self.start(),
-            self.size()
+            self.end(),
```
This can be `self.start() + self.size()`

sdharar replied:
This has been resolved in previous commits.

arastoob commented on src/boxes/mp4_box.rs:
```diff
@@ -55,16 +63,39 @@ impl std::fmt::Debug for dyn Mp4Box {
             }
         }
+
+        let off_print = self.offsets();
```
Having the offset for all the modules added this extra code here. Maybe you can handle this in the `stco` module?

sdharar replied:
This has been resolved in the latest commit.

arastoob commented on src/boxes/media/mod.rs:
```diff
@@ -73,27 +76,18 @@ impl Mp4Box for Media {
-                        index,
-                        level + 1,
-                    )?) as Box<dyn Mp4Box>;
+                    let b =
```
Have you used `cargo fmt` to auto-format the code? There are many formatting-related changes.

sdharar replied:
Yes, that is correct. I think this can be ignored as this is not making changes to the logic of the code?

arastoob commented on src/boxes/media/sample_table.rs:
```diff
@@ -123,85 +125,56 @@ impl Mp4Box for SampleTable {
+                        Box::new(InnerAtom::parse(reader, index, level + 1)?) as Box<dyn Mp4Box>;
                     sample_table.ctts = Some(b);
                 }
-                "stss" => {
```
What happened to this? Why it is removed?

sdharar replied:
It's not removed, it's just re-arranged. See on the right side. This and other re-arrangement are done to print them according to the order specified in the standard ISO document to print them as per them. 

I suppose since this isn't a major change in logic here, can be ignored?

arastoob commented on src/boxes/movie.rs:
```diff
@@ -12,8 +13,8 @@ use std::io::{Read, Seek, SeekFrom};
     mvhd: Option<Box<dyn Mp4Box>>, // movie header
-    tracks: Vec<Box<dyn Mp4Box>>,  // movie tracks
     udta: Option<Box<dyn Mp4Box>>, // user data
+    tracks: Vec<Box<dyn Mp4Box>>,  // movie tracks
```
Any reason to change the order of the fields?

sdharar replied:
The reason is to keep the ease of understanding and keep them in order according to standard parsing fields as per ISO documnet. This is not a major change and can be ignored?

arastoob commented on src/boxes/movie.rs:
```diff
@@ -59,45 +60,41 @@ impl Mp4Box for Movie {
+                                   // println!("index: {}, len: {}", index, len);
 
-        while index < len {
+        while index < start + len {
```
I'm not sure about these changes for all the modules. I know that you need this for `wide`, but not sure about others. You should test this for different files...

sdharar replied:
The reason we are making this change is that the older logic doesn't work on all type of mp4 files. This is not needed for "wide" atom. This is changed because: the older logic of __ index < len __ does not take into consideration the fact that larger atoms such mdat may come in second or third place. It always expects the index to be smaller than the len which may not be the case when you see a larger-sized atom. The older logic will result in buffer not filled related errors. I've tested this on different mp4 files and this appears to work fine.  Let me know if you come across any mp4 file that doesn't support this.

arastoob commented on src/boxes/movie.rs:
```diff
@@ -59,45 +60,41 @@ impl Mp4Box for Movie {
             let name = std::str::from_utf8(&name)?;
+
             let name = AtomName::from(name);
+            // println!("Name: {:?}, Size: {:?} and Index: {:?}", name, size, index);
```
We shouldn't commit the commented code, please remove them

sdharar replied:
Noted.

sdharar replied:
This has been resolved in previous commits.

arastoob commented on src/boxes/movie.rs:
```diff
@@ -124,17 +125,21 @@ impl Mp4Box for Movie {
             fields.push(udta.as_ref());
         }
 
+        for track in self.tracks.iter() {
```
Again, why the order of the code is changed? Any particular reason?

sdharar replied:
See: 
1. https://github.com/musec/fmvparse/pull/3/files/40f4dfc53032079fad8adfa10b9a0881a93e22bb#r872698293
2. https://github.com/musec/fmvparse/pull/3/files/40f4dfc53032079fad8adfa10b9a0881a93e22bb#r872700454
3. https://github.com/musec/fmvparse/pull/3/files/40f4dfc53032079fad8adfa10b9a0881a93e22bb#r872695330


Also, since this doesn't make any change to the code logic, is that a problem?

arastoob commented on src/boxes/wide.rs:
```diff
@@ -0,0 +1,50 @@
+use std::io::{Read, Seek};
+
+#[derive(Debug)]
+pub struct Wide {
```
If this struct doesn't its own methods and fields, why not handle it with `InnerAtom`? We shouldn't add redundant code if we don't actually need

sdharar replied:
I'm following your base logic/structure here, as none of the first level atoms like ftyp, free, mdat have their own methods and fields but are under separate files. Do they also require handling with InnerAtom to avoid redundancy?

arastoob replied:
If the other modules are like this in the first level, so keep this one as well for consistency.

sdharar replied:
I've done this in the newer commit on the stco-parsing branch.

arastoob commented on src/header.rs:
```diff
@@ -5,13 +5,14 @@
 use crate::error::Error;
 use byteorder::{BigEndian, ByteOrder};
-use std::io::{Seek, SeekFrom, Read};
+use std::io::{Read, Seek, SeekFrom};
```
The same line, just different orders?

sdharar replied:
This must have happened as a result of the rust formatter on my Vscode, it tries to keep imports in alphabetical order. This can be ignored.

arastoob commented on src/mp4.rs:
```diff
@@ -13,9 +14,11 @@ use std::path::Path;
     moov: Option<Box<dyn Mp4Box>>, // container for all the metadata
     free: Option<Box<dyn Mp4Box>>, // free space
-    mdat: Option<Box<dyn Mp4Box>>, //  media data container
+    skip: Option<Box<dyn Mp4Box>>, // Unused space available in file
```
Does this need to be a Mp4Box?

sdharar replied:
Yes, as it's a first-level atom (with no further level necessary down the line under itself) just like Free and others.

arastoob replied:
I mean if it's some bytes that are skipped, why do we need to keep them in our structure?

sdharar replied:
These are present in the ISO documentation for Mp4 in the structure itself. I suppose we should follow what's considered "standard" and should be able to claim the software to be compliant with the official mp4 standard in the end? Do we have any issues with following the standard ISO document?

sdharar committed:
(1) Removed end() method and replaced code in mp4box.rs with ___self.start() + self.size() as u64___ to print end address.
(2) Removed redundent code and the rust files => file_type.rs, wide.rs, media_data.rs, free.rs and handled these through code present in inner.rs.
(3) Renamed offset method to getmetadata() for more clarity on the prupose of the method.

arastoob commented on src/boxes/movie.rs:
```diff
@@ -124,17 +119,21 @@ impl Mp4Box for Movie {
     }
 
+    fn getmetadata(&self) -> Option<Vec<u64>> {
+        None
```
If this is None everywhere except one place, it shouldn't be in the trait...

sdharar replied:
This is the only pending issue, which I'm working on based on our in-person discussion. I will post a commit after this, and let you review the changed code.

arastoob commented:
**A general comment**
There are still some issues not addressed yet. However, let's continue with this code as it's working and you shouldn't be blocked, but don't merge it into the master branch.

sdharar committed:
Removed getmatadat()/offset() method and handled Chunk Offset Table parsing in stco.rs

sdharar commented:
> **A general comment** There are still some issues not addressed yet. However, let's continue with this code as it's working, and you shouldn't be blocked, but don't merge it into the master branch.

@arastoob @trombonehero I've made another commit on the `stco-parsing` branch, considering the previous comments. This should address the pending issues proposed by _Dr. Bozorgi_, including handling of `stco` atom parsing in `stco.rs `file instead of having a method in other files. 

**Note:** Unlike previous code in the `master` branch containing some redundant code, lacking parsing for `ChunkOffsetTable` and bugs **when it comes to parsing DJI-based Mp4 video files**, The changed code in the `stco-parsing` branch is now able to parse the necessary metadata for the files captured through _DJI_ drones (including the parsing of stco atom containing offset addresses which is absent the older code present on `master` branch).

**Suggested course of action:** Since the `master`/`main` branch must contain the latest work/changes/upgrades on the CLI/tool, I would suggest merging this branch to `master`/`main` and deleting the older branch. 

Please let me know if you have any issues or concerns with this.

Thanks.
Samir Dharar

sdharar commented:
Merging this branch to master based on our discussion today.
